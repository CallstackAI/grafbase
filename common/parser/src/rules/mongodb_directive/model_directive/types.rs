pub(crate) mod create;
pub(crate) mod delete;
pub(crate) mod filter;
pub(crate) mod generic;
pub(crate) mod oneof;
