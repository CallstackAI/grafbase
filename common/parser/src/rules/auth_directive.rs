use super::visitor::{Visitor, VisitorContext};

use dynaql::ServerError;
use dynaql_parser::types::ConstDirective;
use dynaql_value::ConstValue;

pub const AUTH_DIRECTIVE: &str = "auth";

pub struct AuthDirective;

#[derive(Debug, serde::Serialize, serde::Deserialize)]
struct Auth {
    providers: Vec<AuthProvider>,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "lowercase")]
#[serde(tag = "type")]
#[serde(deny_unknown_fields)]
enum AuthProvider {
    #[serde(rename_all = "camelCase")]
    Oidc {
        issuer: url::Url,
        groups: Option<Vec<String>>,
    },
}

impl<'a> Visitor<'a> for AuthDirective {
    // FIXME: this snippet is parsed, but not enforced by the server, why?
    fn directives(&self) -> String {
        r#"
        directive @auth(providers: [AuthProviderDefinition!]!) on SCHEMA
        input AuthProviderDefinition {
          type: String!
          issuer: String!
        }
        "#
        .to_string()
    }

    fn enter_schema(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        schema_definition: &'a dynaql::Positioned<dynaql_parser::types::SchemaDefinition>,
    ) {
        if let Some(directive) = schema_definition
            .node
            .directives
            .iter()
            .find(|d| d.node.name.node == AUTH_DIRECTIVE)
        {
            match (&directive.node).try_into() as Result<Auth, ServerError> {
                Ok(auth) => {
                    ctx.registry.get_mut().auth = Some(auth.into());
                }
                Err(err) => {
                    ctx.report_error(vec![directive.pos], err.message);
                }
            }
        }
    }
}

impl TryFrom<&ConstDirective> for Auth {
    type Error = ServerError;

    fn try_from(value: &ConstDirective) -> Result<Self, Self::Error> {
        let pos = Some(value.name.pos);

        let arg = match value.get_argument("providers") {
            Some(arg) => match &arg.node {
                ConstValue::List(value) => value,
                _ => return Err(ServerError::new("auth providers must be a list", pos)),
            },
            None => return Err(ServerError::new("auth providers missing", pos)),
        };

        let providers = arg
            .iter()
            .map(AuthProvider::try_from)
            .collect::<Result<_, _>>()
            .map_err(|err| ServerError::new(err.message, pos))?;

        Ok(Auth { providers })
    }
}

impl TryFrom<&ConstValue> for AuthProvider {
    type Error = ServerError;

    fn try_from(value: &ConstValue) -> Result<Self, Self::Error> {
        // We convert the value to JSON to leverage serde for deserialization
        let value = match value {
            ConstValue::Object(_) => value
                .clone()
                .into_json()
                .map_err(|err| ServerError::new(err.to_string(), None))?,
            _ => return Err(ServerError::new("auth provider must be an object", None)),
        };

        let provider: AuthProvider =
            serde_json::from_value(value).map_err(|err| ServerError::new(format!("auth provider: {err}"), None))?;

        Ok(provider)
    }
}

impl From<Auth> for dynaql::Auth {
    fn from(auth: Auth) -> Self {
        Self {
            oidc_providers: auth
                .providers
                .iter()
                .map(|provider| match provider {
                    AuthProvider::Oidc { issuer, groups } => dynaql::OidcProvider {
                        issuer: issuer.clone(),
                        groups: groups.clone(),
                    },
                })
                .collect(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::AuthDirective;
    use crate::rules::visitor::{visit, VisitorContext};
    use dynaql_parser::parse_schema;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_oidc_basic() {
        let schema = r#"
            schema @auth(providers: [
              { type: "oidc", issuer: "https://my.idp.com" }
            ]) {
              query: Boolean # HACK: make top-level auth directive work
            }
            "#;

        let schema = parse_schema(schema).unwrap();
        let mut ctx = VisitorContext::new(&schema);
        visit(&mut AuthDirective, &mut ctx, &schema);

        assert!(ctx.errors.is_empty());
        assert_eq!(
            ctx.registry.borrow().auth.as_ref().unwrap().oidc_providers,
            vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups: None,
            }]
        );
    }

    #[test]
    fn test_oidc_groups() {
        let schema = r#"
            schema @auth(providers: [
              { type: "oidc", issuer: "https://my.idp.com", groups: ["admin"] }
            ]) {
              query: Boolean
            }
            "#;

        let schema = parse_schema(schema).unwrap();
        let mut ctx = VisitorContext::new(&schema);
        visit(&mut AuthDirective, &mut ctx, &schema);

        assert!(ctx.errors.is_empty());
        assert_eq!(
            ctx.registry.borrow().auth.as_ref().unwrap().oidc_providers,
            vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups: Some(vec!["admin".to_string()]),
            }]
        );
    }

    #[test]
    fn test_oidc_missing_field() {
        let schema = r#"
            schema @auth(providers: [
              { type: "oidc" }
            ]) {
              query: Boolean
            }
            "#;

        let schema = parse_schema(schema).unwrap();
        let mut ctx = VisitorContext::new(&schema);
        visit(&mut AuthDirective, &mut ctx, &schema);

        assert_eq!(ctx.errors.len(), 1);
        assert_eq!(
            ctx.errors.get(0).unwrap().message,
            "auth provider: missing field `issuer`",
        );
    }
}
