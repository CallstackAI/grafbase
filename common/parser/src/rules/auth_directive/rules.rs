use std::collections::HashSet;

use serde::{Deserialize, Serialize};

use dynaql::ServerError;
use dynaql_value::ConstValue;

use super::operations::Operations;
use crate::VisitorContext;

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(tag = "allow")]
#[serde(deny_unknown_fields)]
#[non_exhaustive]
pub enum AuthRule {
    /// Public data access via API keys
    // Ex: { allow: anonymous }
    #[serde(alias = "public")]
    #[serde(rename_all = "camelCase")]
    Anonymous {
        // Note: we don't support operations as our playground needs full access
    },

    /// Signed-in user data access via OIDC
    // Ex: { allow: private }
    //     { allow: private, operations: [create, read] }
    #[serde(rename_all = "camelCase")]
    Private {
        #[serde(default)]
        operations: Operations,
    },

    /// User group-based data access via OIDC
    // Ex: { allow: groups, groups: ["admin"] }
    //     { allow: groups, groups: ["admin"], operations: [update, delete] }
    #[serde(rename_all = "camelCase")]
    Groups {
        #[serde(with = "::serde_with::rust::sets_duplicate_value_is_error")]
        groups: HashSet<String>,

        #[serde(default)]
        operations: Operations,
    },

    /// Owner-based data access via OIDC
    // Ex: { allow: owner }
    //     { allow: owner, operations: [create, read] }
    #[serde(rename_all = "camelCase")]
    Owner {
        #[serde(default)]
        operations: Operations,
    },
}

impl AuthRule {
    pub fn from_value(_ctx: &VisitorContext<'_>, value: &ConstValue) -> Result<Self, ServerError> {
        // We convert the value to JSON to leverage serde for deserialization
        let value = match value {
            ConstValue::Object(_) => value
                .clone()
                .into_json()
                .map_err(|err| ServerError::new(err.to_string(), None))?,
            _ => return Err(ServerError::new("auth rule must be an object", None)),
        };

        let rule: AuthRule =
            serde_json::from_value(value).map_err(|err| ServerError::new(format!("auth rule: {err}"), None))?;

        Ok(rule)
    }
}
