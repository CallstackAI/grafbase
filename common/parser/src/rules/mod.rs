pub mod basic_type;
pub mod check_type_validity;
pub mod check_types_underscore;
pub mod model_directive;
pub mod visitor;
