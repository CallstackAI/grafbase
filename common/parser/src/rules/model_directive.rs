//! Implement the model directive
//!
//! When a @model directive is present for a type, we generate the associated type into the
//! registry and generate the CRUDL configuration for this type.
//!
//! Flow:
//!  -> When there is a @model directive on a type
//!  -> Must be an ObjectType
//!  -> Must have primitives
//!  -> Must have a non_nullable ID type
//!
//! Then:
//!  -> Create the ObjectType
//!  -> Create the ReadById Query
//!  -> Create the Create Mutation
//!
//! TODO: Should have either: an ID or a PK

use super::relations::generate_metarelation;
use super::visitor::{Visitor, VisitorContext};
use crate::registry::add_list_query_paginated;
use crate::registry::add_remove_query;
use crate::registry::{add_create_mutation, add_update_mutation};
use crate::utils::is_modelized_node;
use crate::utils::to_base_type_str;
use crate::utils::to_lower_camelcase;
use crate::utils::{is_id_type_and_non_nullable, pagination_arguments};
use case::CaseExt;
use dynaql::indexmap::IndexMap;
use dynaql::registry::resolvers::context_data::ContextDataResolver;
use dynaql::registry::resolvers::dynamo_querying::DynamoResolver;
use dynaql::registry::{is_array_basic_type, MetaType};
use dynaql::registry::{
    resolvers::Resolver, resolvers::ResolverType, transformers::Transformer, variables::VariableResolveDefinition,
};
use dynaql::registry::{Constraint, MetaField};
use dynaql::registry::{ConstraintType, MetaInputValue};
use dynaql::Operations;
use dynaql_parser::types::{Type, TypeKind};
use if_chain::if_chain;

pub struct ModelDirective;

pub const MODEL_DIRECTIVE: &str = "model";
pub const UNIQUE_DIRECTIVE: &str = "unique";

impl<'a> Visitor<'a> for ModelDirective {
    fn directives(&self) -> String {
        r#"
        directive @model on OBJECT
        "#
        .to_string()
    }

    fn enter_type_definition(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        type_definition: &'a dynaql::Positioned<dynaql_parser::types::TypeDefinition>,
    ) {
        let directives = &type_definition.node.directives;
        if_chain! {
            if directives.iter().any(|directive| directive.node.name.node == MODEL_DIRECTIVE);
            if let TypeKind::Object(object) = &type_definition.node.kind;
            if let Some(id_field) = object.fields.iter().find(|x| is_id_type_and_non_nullable(&x.node));
            then {
                let type_name = type_definition.node.name.node.to_string();
                let mut connection_edges = Vec::new();
                // If it's a modeled Type, we create the associated type into the registry.
                // Without more data, we infer it's from our modelization.
                ctx.registry.get_mut().create_type(&mut |_| MetaType::Object {
                    name: type_name.clone(),
                    description: type_definition.node.description.clone().map(|x| x.node),
                    fields: {
                        let mut fields = IndexMap::new();
                        for field in &object.fields {
                            let name = field.node.name.node.to_string();
                            let ty = is_modelized_node(&ctx.types, &field.node.ty.node);
                            let relation = ty.map(|_ty| {
                                generate_metarelation(&type_definition.node, &field.node)
                            });

                            let is_edge = relation.as_ref().map(|x| {
                                let target_ty = Type::new(x.relation.1.as_str()).expect("shouldn't fail");
                                is_modelized_node(&ctx.types, &target_ty).is_some()
                            }).unwrap_or_default();

                            let transforms = if is_edge {
                                None
                            } else {
                                Some(vec![Transformer::DynamoSelect {
                                    property: if name == "id" {
                                        "__sk".to_string()
                                    } else {
                                        name.clone()
                                    },
                                }])
                            };

                            let is_expecting_array = is_array_basic_type(&field.node.ty.to_string());
                            let relation_array = relation.is_some() && is_expecting_array;

                            let resolve = match (is_edge, &relation) {
                                (true, Some(relation)) => Some(Resolver {
                                    id: Some(format!("{}_edge_resolver", relation.relation.0.to_lowercase())),
                                    r#type: ResolverType::ContextDataResolver(if relation_array {
                                        ContextDataResolver::EdgeArray {
                                            key: relation.relation.0.clone(),
                                            relation_name: relation.name.clone(),
                                            expected_ty: to_base_type_str(&field.node.ty.node.base),
                                        }
                                    } else {
                                        ContextDataResolver::SingleEdge {
                                            key: relation.relation.0.clone(),
                                            relation_name: relation.name.clone(),
                                        }
                                    }),
                                }),
                                (false, Some(relation)) => Some(Resolver {
                                    id: None,
                                    r#type: ResolverType::ContextDataResolver(ContextDataResolver::LocalKey { key: relation.name.clone() }),
                                }),
                                (false, None) => Some(Resolver {
                                    id: None,
                                    r#type: ResolverType::ContextDataResolver(ContextDataResolver::LocalKey { key: type_name.to_string() }),
                                }),
                                _ => unreachable!("Can't happen yet"),
                            };

                            fields.insert(name.clone(), MetaField {
                                name,
                                description: field.node.description.clone().map(|x| x.node),
                                args: if relation_array {
                                    pagination_arguments()
                                } else {
                                    Default::default()
                                },
                                ty: if relation_array {
                                    format!("{}Connection", to_base_type_str(&field.node.ty.node.base).to_camel())
                                } else {
                                    field.node.ty.clone().node.to_string()
                                },
                                deprecation: Default::default(),
                                cache_control: Default::default(),
                                external: false,
                                requires: None,
                                provides: None,
                                visible: None,
                                compute_complexity: None,
                                resolve,
                                edges: if is_edge {
                                    let edge_type = to_base_type_str(&field.node.ty.node.base);
                                    connection_edges.push(edge_type.clone());
                                    vec![edge_type]
                                } else {
                                    Vec::new()
                                },
                                relation,
                                transforms,
                                required_operation: None,
                            });
                        };
                        fields
                    },
                    cache_control: dynaql::CacheControl {
                        public: true,
                        max_age: 0usize,
                    },
                    extends: false,
                    keys: None,
                    visible: None,
                    is_subscription: false,
                    is_node: true,
                    rust_typename: type_name.clone(),
                    constraints: object.fields
                        .iter()
                        .filter_map(|field| field
                            .node
                            .directives
                            .iter()
                            .find(|directive| directive.node.name.node == UNIQUE_DIRECTIVE)
                            .map(|_| Constraint {
                                field: field.node.name.to_string(),
                                r#type: ConstraintType::Unique,
                            }))
                        .collect(),
                }, &type_name, &type_name);

                ctx.queries.push(MetaField {
                    // byID query
                    name: to_lower_camelcase(&type_name),
                    description: Some(format!("Get {} by ID", type_name)),
                    args: {
                        let mut args = IndexMap::new();
                        args.insert(
                            "id".to_owned(),
                            MetaInputValue {
                                name: "id".to_owned(),
                                ty: "ID!".to_string(),
                                visible: None,
                                description: Some(format!("{} id", type_name)),
                                is_secret: false,
                                default_value: None,
                            },
                        );
                        args
                    },
                    ty: type_name.clone(),
                    deprecation: dynaql::registry::Deprecation::NoDeprecated,
                    cache_control: dynaql::CacheControl {
                        public: true,
                        max_age: 0usize,
                    },
                    external: false,
                    provides: None,
                    requires: None,
                    visible: None,
                    compute_complexity: None,
                    edges: Vec::new(),
                    relation: None,
                    resolve: Some(Resolver {
                        id: Some(format!("{}_resolver", type_name.to_lowercase())),
                        // TODO: Should be defined as a ResolveNode
                        // Single entity
                        r#type: ResolverType::DynamoResolver(DynamoResolver::QueryPKSK {
                            pk: VariableResolveDefinition::InputTypeName("id".to_owned()),
                            sk: VariableResolveDefinition::InputTypeName("id".to_owned()),
                        })
                    }),
                    transforms: None,
                    required_operation: Some(Operations::GET),
                });

                add_create_mutation(ctx, &type_definition.node, object, &type_name);
                add_update_mutation(ctx, &type_definition.node, object, &type_name);

                add_list_query_paginated(ctx, &type_name, connection_edges);
                add_remove_query(ctx, &id_field.node, &type_name)
            }
        }
    }
}
