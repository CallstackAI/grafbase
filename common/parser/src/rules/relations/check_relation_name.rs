//! ### What it does
//!
//! Check that relation names only contain valid characters.
//!
//! ### Why?
//!
//! There are limitations on these and if we don't validate here
//! users get runtime errors

use super::relations_engine::{NAME_ARGUMENT, RELATION_DIRECTIVE};
use crate::rules::visitor::{Visitor, VisitorContext};
use dynaql_value::ConstValue;
use if_chain::if_chain;
use regex::Regex;

pub struct CheckRelationName;

static NAME_CHARS: &str = r"[_a-zA-Z0-9]";

lazy_static::lazy_static! {
    static ref NAME_RE: Regex = Regex::new(&format!("^{NAME_CHARS}*$")).unwrap();
}

impl<'a> Visitor<'a> for CheckRelationName {
    fn enter_directive(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        directive: &'a dynaql::Positioned<dynaql_parser::types::ConstDirective>,
    ) {
        if_chain! {
            let directive = &directive.node;
            if directive.name.node == RELATION_DIRECTIVE;
            if let Some(value) = directive.get_argument(NAME_ARGUMENT);
            if let ConstValue::String(name) = &value.node;
            if !NAME_RE.is_match(name);
            then {
                ctx.report_error(
                    vec![value.pos],
                    format!("Relation names should only contain {NAME_CHARS} but {name} does not"),
                );
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::CheckRelationName;
    use crate::rules::visitor::{visit, VisitorContext};
    use dynaql_parser::parse_schema;

    #[test]
    fn should_error_on_invalid_name() {
        let schema = r#"
            type Todo @model {
                secondaryAuthors: [Author] @relation(name: "second-author")
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckRelationName, &mut ctx, &schema);

        assert!(!ctx.errors.is_empty(), "shouldn't be empty");
        assert_eq!(ctx.errors.len(), 1, "should have one error");
        assert_eq!(
            ctx.errors.get(0).unwrap().message,
            "Relation names should only contain [_a-zA-Z0-9] but second-author does not",
        );
    }

    #[test]
    fn should_allow_valid_names() {
        let schema = r#"
            type Todo @model {
                secondaryAuthors: [Author] @relation(name: "second_author")
                evenMoreAuthors: [Author] @relation(name: "moreAuthors")
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckRelationName, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should not have any error");
    }
}
