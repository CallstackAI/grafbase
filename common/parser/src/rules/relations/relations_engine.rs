//! Implement the Relation Engine
use crate::rules::model_directive::MODEL_DIRECTIVE;
use crate::utils::is_modelized_node;
use crate::{Visitor, VisitorContext};
use async_graphql::indexmap::map::Entry;
use async_graphql::registry::relations::MetaRelation;
use async_graphql::Value;
use async_graphql_parser::types::{FieldDefinition, Type, TypeDefinition, TypeKind};
use if_chain::if_chain;

/// Implement the Relation Engine
///
/// We need to define the relation before hand, to do that we have two mecanism
/// working to define relation:
///
/// - Implicit: By having an explicit relation on two modelized node.
/// - Explicit: By having the relation defined by the `@relation` directive
///
/// A relation can only exist between two nodes.
///
/// # Algorithm
///
/// For each modelized node, we go into each fields, for each field:
///
/// -> We pass into the field
/// --> If modelized
/// --> Attribute a relation name based on those two types.
/// --> Store it into the VisitorContext (Will be used to compare between two iteration of a schema
/// if there is a change in relations)
/// --> (Store it into the generated type, need dynaql work)
pub struct RelationEngine;

pub const RELATION_DIRECTIVE: &str = "relation";

/// Generate a `MetaRelation` if possible
pub fn generate_metarelation(ty: &TypeDefinition, field: &FieldDefinition) -> MetaRelation {
    let type_name = ty.name.node.to_string();
    let name = field
        .directives
        .iter()
        .find(|directive| directive.node.name.node == RELATION_DIRECTIVE)
        .and_then(|dir| dir.node.get_argument("name"))
        .and_then(|name| match &name.node {
            Value::String(inner) => Some(inner.clone()),
            _ => None,
        });

    MetaRelation::new(name, &Type::new(&type_name).expect("Shouldn't fail"), &field.ty.node)
}

impl<'a> Visitor<'a> for RelationEngine {
    fn directives(&self) -> String {
        r#"
        directive @relation(
          """
          The name of the relation
          """
          name: String!
        ) on FIELD_DEFINITION
        "#
        .to_string()
    }

    fn enter_type_definition(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        type_definition: &'a async_graphql::Positioned<async_graphql_parser::types::TypeDefinition>,
    ) {
        let directives = &type_definition.node.directives;
        if_chain! {
            // We do check if it's a modelized node
            // TODO: Create an abstraction over it
            let _ = directives.iter().find(|directive| directive.node.name.node == MODEL_DIRECTIVE);
            if let TypeKind::Object(object) = &type_definition.node.kind;
            // We do check if it's a modelized node
            then {
                // We iterate over fields that reprensent a relation to check than
                let mut errors = Vec::new();
                for (field, _) in object.fields.iter().filter_map(|field| is_modelized_node(&ctx.types, &field.node.ty.node).map(|ty| (field, ty))) {
                    let relation = generate_metarelation(&type_definition.node, &field.node);
                    match ctx.relations.entry(relation.name.clone()) {
                        Entry::Vacant(vac) => {
                            vac.insert(relation);
                        }
                        Entry::Occupied(mut oqp) => {
                            if let Err(err) = oqp.get_mut().with(relation) {
                                errors.push((field.pos, err));
                            }
                        }
                    };

                }

                for (pos, err) in errors {
                    ctx.report_error(
                    vec![pos],
                    format!("Relations issues: {}", err),
                );
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::RelationEngine;
    use crate::rules::visitor::{visit, VisitorContext};
    use async_graphql_parser::parse_schema;
    use insta::assert_debug_snapshot;
    use serde_json as _;

    #[test]
    fn one_to_one_relation_monodirectional() {
        let schema = r#"
            type Author @model {
                id: ID!
            }

            type Post @model {
                id: ID!
                publishedBy: Author
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut RelationEngine, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should be empty");
        assert_eq!(ctx.relations.len(), 1 as usize, "Should have only one relation");
        assert_eq!(
            ctx.relations.iter().next().unwrap().1.birectional,
            false,
            "Should be monodirectional"
        );
        assert_debug_snapshot!(&ctx.relations);
    }

    #[test]
    fn one_to_one_relation_bidirectionnal() {
        let schema = r#"
            type Author @model {
                id: ID!
                published: Post
            }

            type Post @model {
                id: ID!
                publishedBy: Author
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut RelationEngine, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should be empty");
        assert_eq!(ctx.relations.len(), 1 as usize, "Should have only one relation");
        assert_eq!(
            ctx.relations.iter().next().unwrap().1.birectional,
            true,
            "Should be bidirectional"
        );
        assert_debug_snapshot!(&ctx.relations);
    }

    #[test]
    fn one_to_many_relation_monodirectional_1() {
        let schema = r#"
            type Author @model {
                id: ID!
            }

            type Post @model {
                id: ID!
                publishedBy: [Author]
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut RelationEngine, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should be empty");
        assert_eq!(ctx.relations.len(), 1 as usize, "Should have only one relation");
        assert_eq!(
            ctx.relations.iter().next().unwrap().1.birectional,
            false,
            "Should be monodirectional"
        );
        assert_debug_snapshot!(&ctx.relations);
    }

    #[test]
    fn one_to_many_relation_monodirectional_2() {
        let schema = r#"
            type Author @model {
                id: ID!
                posts: [Post]
            }

            type Post @model {
                id: ID!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut RelationEngine, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should be empty");
        assert_eq!(ctx.relations.len(), 1 as usize, "Should have only one relation");
        assert_eq!(
            ctx.relations.iter().next().unwrap().1.birectional,
            false,
            "Should be monodirectional"
        );
        assert_debug_snapshot!(&ctx.relations);
    }

    #[test]
    fn one_to_many_relation_bidirectional_1() {
        let schema = r#"
            type Author @model {
                id: ID!
                post: Post!
            }

            type Post @model {
                id: ID!
                publishedBy: [Author]
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut RelationEngine, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should be empty");
        assert_eq!(ctx.relations.len(), 1 as usize, "Should have only one relation");
        assert_eq!(
            ctx.relations.iter().next().unwrap().1.birectional,
            true,
            "Should be bidirectional"
        );
        assert_debug_snapshot!(&ctx.relations);
    }

    #[test]
    fn one_to_many_relation_bidirectional_2() {
        let schema = r#"
            type Author @model {
                id: ID!
                posts: [Post]
            }

            type Post @model {
                id: ID!
                author: Author!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut RelationEngine, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should be empty");
        assert_eq!(ctx.relations.len(), 1 as usize, "Should have only one relation");
        assert_eq!(
            ctx.relations.iter().next().unwrap().1.birectional,
            true,
            "Should be bidirectional"
        );
        assert_debug_snapshot!(&ctx.relations);
    }

    #[test]
    fn many_to_many_relation_monodirectional() {
        let schema = r#"
            type Author @model {
                id: ID!
                posts: [Post!]
            }

            type Post @model {
                id: ID!
                publishedBy: [Author!]
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut RelationEngine, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should be empty");
        assert_eq!(ctx.relations.len(), 1 as usize, "Should have only one relation");
        assert_eq!(
            ctx.relations.iter().next().unwrap().1.birectional,
            true,
            "Should be bidirectional"
        );
        assert_debug_snapshot!(&ctx.relations);
    }
}
