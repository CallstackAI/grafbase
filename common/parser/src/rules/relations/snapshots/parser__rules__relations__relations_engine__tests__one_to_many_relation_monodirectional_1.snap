---
source: common/parser/src/rules/relations/relations_engine.rs
expression: "&ctx.relations"
---
{
    "AuthorRelatePost": MetaRelation {
        name: "AuthorRelatePost",
        kind: OneToMany,
        relation: (
            Some(
                "Post",
            ),
            "Author",
        ),
        birectional: false,
    },
}
